pkg/agent/core/metrics.go

```diff
@@ -0,0 +1,182 @@
+package core
+
+// Abstraction over the []metrics -> CU "algorithm"
+//
+// In general, there's only *really* one ScalingAlgorithm implementation, but we keep it generic so
+// that tests can dependency-inject an alternate implementation that's more
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/neondatabase/autoscaling/pkg/api"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type Metrics struct {
+	LoadAverage1Min  float32
+	LoadAverage5Min  float32 // unused. To be removed when api.Metrics.LoadAverage5Min is removed.
+	MemoryUsageBytes float32
+}
+
+// ToAPI converts the Metrics to the subset used for communications with the scheduler plugin and
+// exposed in pkg/api
+//
+// ToAPI implements ToAPIMetrics.
+func (m Metrics) ToAPI() api.Metrics {
+	return api.Metrics{
+		LoadAverage1Min:  m.LoadAverage1Min,
+		LoadAverage5Min:  m.LoadAverage5Min,
+		MemoryUsageBytes: m.MemoryUsageBytes,
+	}
+}
+
+// ScalingAlgorithm is the abstract interface provided by the default scaling algorithm, so that a
+// simpler version can be dependency-injected for easier testing.
+//
+// And also, separation of concerns is nice. Where state.go (and State within it) are more concerned
+// with the overall process of scaling approval and notification by communicating with other
+// components, the implementation(s) of ScalingAlgorithm can be more focused, making the calculation
+// of desired state abundantly clear.
+type ScalingAlgorithm[M ToAPIMetrics] interface {
+	// Update adds recent metrics
+	Update(metrics M)
+
+	// GoalCU returns the desired number of compute units that the VM should be using, given data
+	// from prior calls to Update and the configured resources equal to one compute unit.
+	GoalCU(scalingConfig api.ScalingConfig, computeUnit api.Resources) (_ uint32, ok bool)
+
+	// DeepCopy produces a copy of the value, intended for use when serializing State
+	DeepCopy() ScalingAlgorithm[M]
+}
+
+// ToApiMetrics is the interface that ScalingAlgorithm's metrics must satisfy - namely, that they
```

Comment:
Why do we need to have multiple implementations of `Metrics`? Can't we have a large struct and put there everything that is needed by any of `ScalingAlgorithm`s?

Reply:
From the PR description:

> this PR switches `pkg/agent/core/state_test.go` to mostly use a new `ScalingAlgorithm` implementation that uses the desired CU as the associated "metrics". This should allow substantial changes to goal CU calculation without needing to change complex, unrelated tests.

The basic idea is that we have a lot of complex tests in `state_test.go` that don't _really_ care about the actual metrics, just the implications on goal CU. So it's better to just pass the goal CU itself as the "metric" so we can make changes to our scaling algorithm without needing to change all the tests.

Reply:
To me metrics is a data object rather an entity, so I agree with @Omrigan that I can't easily see a case where having metrics as an interface can be useful. Also, I didn't quite get that why can't we use dummy values for metrics in tests.